source/validate_layout.cpp

```diff
@@ -0,0 +1,205 @@
+}
+
+// Function declaration validation is performed by making sure that the
+// FunctionParameter and FunctionEnd instructions only appear inside of
+// functions. It also ensures that the Function instruction does not appear
+// inside of another function. This stage ends when the first label is
+// encountered inside of a function.
+spv_result_t FunctionScopedInstructions(ValidationState_t& _,
+                                        const spv_parsed_instruction_t* inst,
+                                        SpvOp opcode) {
+  if (_.isOpcodeInCurrentLayoutSection(opcode)) {
+    switch (opcode) {
+      case SpvOpFunction:
+        if (_.in_function_body()) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << "Cannot declare a function in a function body";
+        }
+        spvCheckReturn(_.get_functions().RegisterFunction(
+            inst->result_id, inst->type_id,
+            inst->words[inst->operands[2].offset],
+            inst->words[inst->operands[3].offset]));
+        if (_.getLayoutSection() == kLayoutFunctionDefinitions)
+          spvCheckReturn(_.get_functions().RegisterSetFunctionDeclType(
+              FunctionDecl::kFunctionDeclDefinition));
+        break;
+
+      case SpvOpFunctionParameter:
+        if (_.in_function_body() == false) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT) << "Function parameter "
+                                                     "instructions must be in "
+                                                     "a function body";
+        }
+        if (_.get_functions().get_block_count() != 0) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << "Function parameters must only appear immediatly after the "
+                    "function definition";
+        }
+        spvCheckReturn(_.get_functions().RegisterFunctionParameter(
+            inst->result_id, inst->type_id));
+        break;
+
+      case SpvOpFunctionEnd:
+        if (_.in_function_body() == false) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << "Function end instructions must be in a function body";
+        }
+        if (_.in_block()) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << "Function end cannot be called in blocks";
+        }
+        if (_.get_functions().get_block_count() == 0 &&
+            _.getLayoutSection() == kLayoutFunctionDefinitions) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT) << "Function declarations "
+                                                     "must appear before "
+                                                     "function definitions.";
+        }
+        spvCheckReturn(_.get_functions().RegisterFunctionEnd());
+        if (_.getLayoutSection() == kLayoutFunctionDeclarations) {
+          spvCheckReturn(_.get_functions().RegisterSetFunctionDeclType(
+              FunctionDecl::kFunctionDeclDeclaration));
+        }
+        break;
+
+      case SpvOpLine:  // ??
```

Comment:
Yes. `OpLine` is handled in `isOpcodeInCurrentLayoutSection` (see [122](https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/validate_types.cpp#L122)) but it seems I have missed `OpNoLine`.